crates/vm/src/vm.rs

```diff
@@ -135,4 +137,172 @@ impl Vm {
+        state_read: &S,
+        op_gas_cost: &impl OpGasCost,
+        gas_limit: GasLimit,
+    ) -> Result<Gas, ExecError<S::Error>>
+    where
+        S: StateReadSync,
+        I: IntoIterator<Item = u8>,
+        I::IntoIter: Unpin,
+    {
+        let bytecode_lazy = BytecodeMappedLazy::new(bytecode_iter);
+        self.exec_sync(access, state_read, bytecode_lazy, op_gas_cost, gas_limit)
+    }
+
+    /// Execute the given operations synchronously from the current state of the VM.
+    ///
+    /// This function uses synchronous state reading and is intended for use
+    /// with in-memory state implementations.
+    ///
+    /// Upon reaching a `Halt` operation or reaching the end of the operation
+    /// sequence, returns the gas spent and the `Vm` will be left in the
+    /// resulting state.
+    pub fn exec_sync<S, OA>(
+        &mut self,
+        access: Access<'_>,
+        state_read: &S,
+        mut op_access: OA,
+        op_gas_cost: &impl OpGasCost,
+        gas_limit: GasLimit,
+    ) -> Result<Gas, ExecError<S::Error>>
+    where
+        S: StateReadSync,
+        OA: OpAccess<Op = Op>,
+        OA::Error: Into<OpError<S::Error>>,
+    {
+        // Track the gas spent.
+        let mut gas_spent: u64 = 0;
+
+        // Execute each operation
+        while let Some(res) = op_access.op_access(self.pc) {
+            let op = res.map_err(|err| ExecError(self.pc, err.into()))?;
+
+            // Calculate the gas cost of the operation.
+            let op_gas = op_gas_cost.op_gas_cost(&op);
+
+            // Check that the operation wouldn't exceed gas limit.
+            let next_spent = gas_spent
+                .checked_add(op_gas)
+                .filter(|&spent| spent <= gas_limit.total)
+                .ok_or(ExecError(
+                    self.pc,
+                    OutOfGasError {
+                        spent: gas_spent,
+                        op_gas,
+                        limit: gas_limit.total,
+                    }
+                    .into(),
+                ))?;
+
+            // Update the gas spent.
+            gas_spent = next_spent;
+
+            // Execute the operation.
+            let res = step_any_op(access, op, self, state_read);
+
+            #[cfg(feature = "tracing")]
+            crate::trace_op_res(
+                &mut op_access,
+                self.pc,
+                &self.stack,
+                &self.memory,
+                res.as_ref(),
+            );
+
+            // Handle the result of the operation.
+            let update = match res {
+                Ok(update) => update,
+                Err(err) => return Err(ExecError(self.pc, err)),
+            };
+
+            // Update the program counter.
+            match update {
+                Some(ProgramControlFlow::Pc(new_pc)) => self.pc = new_pc,
+                Some(ProgramControlFlow::Halt) => break,
+                None => self.pc += 1,
+            }
+        }
+        Ok(gas_spent)
+    }
 }
```

Comment:
Maybe we should replace the old `sync::exec` function with this one that tracks gas, rather than keep both around?

Reply:
Might do this is the follow up remove async PR but yeh good idea